ibis/expr/operations.py

```diff
@@ -75,32 +76,33 @@ def flat_args(self):
             else:
                 yield arg
 
+    def __hash__(self):
+        if not hasattr(self, '_hash'):
+            self._hash = hash(
+                (type(self),) + tuple(
+                    element.op() if isinstance(element, ir.Expr) else element
+                    for element in self.flat_args()
```

Comment:
We should get rid of `flat_args` and prevent having any non-expressions in the hierarchy, except the literal leaves.